lib/cloud/azure/kubernetes.go

```diff
@@ -0,0 +1,688 @@
+	if cfg.ExecProvider != nil {
+		return nil, trace.BadParameter("exec auth format not supported")
+	}
+	return cfg, nil
+}
+
+// AKSClusterFromManagedCluster converts an Azure armcontainerservice.ManagedCluster into AKSCluster.
+func AKSClusterFromManagedCluster(cluster *armcontainerservice.ManagedCluster) (*AKSCluster, error) {
+	result := &AKSCluster{
+		Name:     StringVal(cluster.Name),
+		Location: StringVal(cluster.Location),
+		Tags:     ConvertTags(cluster.Tags),
+	}
+	if cluster.Identity != nil {
+		result.TenantID = StringVal(cluster.Identity.TenantID)
+	}
+	if subID, groupName, err := extractSubscriptionAndGroupName(cluster.ID); err == nil {
+		result.GroupName, result.SubscriptionID = groupName, subID
+	}
+
+	if cluster.Properties == nil {
+		return nil, trace.BadParameter("invalid AKS Cluster Properties")
+	}
+
+	if !isAKSClusterRunning(cluster.Properties) {
+		return nil, trace.BadParameter("AKS cluster not running")
+	}
+
+	if cluster.Properties.AADProfile != nil && ptrToVal(cluster.Properties.AADProfile.EnableAzureRBAC) {
+		result.Properties = AKSClusterProperties{
+			AccessConfig: AzureRBAC,
+		}
+	} else if cluster.Properties.AADProfile != nil {
+		result.Properties = AKSClusterProperties{
+			AccessConfig:  AzureAD,
+			LocalAccounts: !ptrToVal(cluster.Properties.DisableLocalAccounts),
+		}
+	} else {
+		result.Properties = AKSClusterProperties{
+			AccessConfig:  LocalAccounts,
+			LocalAccounts: true,
+		}
+	}
+
+	return result, nil
+}
+
+func ptrToVal[T any](ptr *T) T {
+	var result T
+	if ptr != nil {
+		result = *ptr
+	}
+	return result
+}
+
+// extractSubscriptionAndGroupName extracts the group name and subscription id from resource id.
+// ids are in the form of:
+// /subscriptions/{subscription_id}/resourcegroups/{resource_group}/providers/Microsoft.ContainerService/managedClusters/{name}
+func extractSubscriptionAndGroupName(id *string) (string, string, error) {
+	if id == nil {
+		return "", "", trace.BadParameter("invalid resource_id provided")
+	}
+	resource, err := armazcore.ParseResourceID(*id)
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+	return resource.SubscriptionID, resource.ResourceGroupName, nil
+}
+
+// azureGroupClaims the configuration settings of the Azure Active Directory allowed principals.
+type azureGroupClaims struct {
+	// Groups - The list of the allowed groups.
+	Groups []string `json:"groups,omitempty"`
+}
+
+func (c *azureGroupClaims) Valid() error {
+	if len(c.Groups) == 0 {
+		return trace.BadParameter("invalid claims received")
+	}
+	return nil
+}
+
+func isAKSClusterRunning(properties *armcontainerservice.ManagedClusterProperties) bool {
+	if properties.PowerState != nil && properties.PowerState.Code != nil &&
+		*properties.PowerState.Code == armcontainerservice.CodeRunning {
+		return true
+	}
+	return false
+}
+
+const (
+	clusterRoleTemplate = `
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: teleport-role
```

Comment:
Nit: Why not just call it "teleport"? "-role" suffix seems redundant cause it's a cluster role resource.